phoenix-core/src/main/java/org/apache/phoenix/iterate/NonAggregateRegionScannerFactory.java

```diff
@@ -109,13 +113,27 @@ public RegionScanner getRegionScanner(final Scan scan, final RegionScanner s) th
         PTable.QualifierEncodingScheme encodingScheme = EncodedColumnsUtil.getQualifierEncodingScheme(scan);
         boolean useNewValueColumnQualifier = EncodedColumnsUtil.useNewValueColumnQualifier(scan);
 
-        Set<KeyValueColumnExpression> arrayKVRefs = Sets.newHashSet();
+        Set<KeyValueColumnExpression> serverParsedKVRefs = Sets.newHashSet();
         KeyValueSchema kvSchema = null;
         ValueBitSet kvSchemaBitSet = null;
-        Expression[] arrayFuncRefs = deserializeArrayPositionalExpressionInfoFromScan(scan, innerScanner, arrayKVRefs);
-        if (arrayFuncRefs != null) {
-            KeyValueSchema.KeyValueSchemaBuilder builder = new KeyValueSchema.KeyValueSchemaBuilder(0);
-            for (Expression expression : arrayFuncRefs) {
+        Expression[]
+                serverParsedArrayFuncRefs = null;
+        if (scan.getAttribute(BaseScannerRegionObserver.SPECIFIC_ARRAY_INDEX) != null) {
+            serverParsedArrayFuncRefs =
+                    deserializeServerParsedPositionalExpressionInfoFromScan(scan,
+                            BaseScannerRegionObserver.SPECIFIC_ARRAY_INDEX, serverParsedKVRefs);
+        }
+        List<Expression> resultList = new ArrayList<>();
+        if (serverParsedArrayFuncRefs != null) {
+            Collections.addAll(resultList, serverParsedArrayFuncRefs);
+        }
+        extractJsonSpecificFuncRefs(scan, serverParsedKVRefs, resultList);
+        Expression[] serverParsedFuncRefs = resultList.toArray(new Expression[0]);
```

Comment:
Maybe you could combine all these into a single function which extracts all the func refs. That will make the code more readable.

Reply:
Just like you have a unified function `deserializeServerParsedPositionalExpressionInfoFromScan`

Reply:
agree. my initial thought was only extracting json specific things out without changing the other code structure. will do this.